modules/gapi/src/compiler/passes/pattern_matching.cpp

```diff
@@ -171,14 +174,20 @@ std::size_t labelOf (const ade::NodeHandle& node, // reader node
     }
 };
 
-inline bool IS_STARTPOINT(const ade::NodeHandle& nh){
+// FIXME: it seems safe for now to check whether nh is INPUT or OUTPUT* through in/out edges.
+//        nonetheless, it's probably better to check via Storage, but current approach is faster
+inline bool IS_STARTPOINT(const ade::NodeHandle& nh) {
     return nh->inEdges().empty();
 }
 
-inline bool IS_ENDPOINT(const ade::NodeHandle& nh){
-    // FIXME: Because of new changes which introduce existence of unused DATA nodes
-    // Try to rely on the nh Data::Storage::OUTPUT
-    return nh->outEdges().empty();
+// * OUTPUT must be explicitly checked via storage due to __unused__ pseudo-out nodes in graph
+inline bool IS_ENDPOINT(const cv::gimpl::GModel::Graph& graph, const ade::NodeHandle& nh) {
+    bool is_endpoint = nh->outEdges().empty();
+    const auto& meta = graph.metadata(nh);
+    if (meta.get<cv::gimpl::NodeType>().t == cv::gimpl::NodeType::DATA) {
+        is_endpoint &= (meta.get<cv::gimpl::Data>().storage == cv::gimpl::Data::Storage::OUTPUT);
+    }
```

Comment:
oh, that's a good question! I'm not sure that it's even possible but if yes, well, I wasn't prepared for this :smile:
(a) && (b) give us __false__ together. But it seems correct that ENDPOINT, so the proper logic is to __always__ just check the storage type value (I wonder: can something similar be done for input nodes as well?):
1. is a DATA node
2. has `Storage::OUTPUT`

@AsyaPronina what was your original idea with `IS_ENDPOINT` and `outEdges().empty()` then?? it seems like current `master` code won't work with @dmatveev 's idea

Reply:
Hi Dmitry, thanks for the pointing this out! Yes, it is a problem, because pattern matching algorithm hopes that output protocol nodes have no output edges.
Hi Andrey, thank you for the comment! Yes, you are right current master is not prepared for this:) 
WIll think about possible options

Reply:
So what did you decide on this? Is it addressed?